R/router.R

```diff
@@ -107,16 +130,105 @@ make_router <- function(default, ...) {
 #' ))
 #' @export
 router_ui <- function() {
-  shiny::tagList(
-    shinyjs::useShinyjs(),
+  shiny::addResourcePath(
+    "shiny.router",
+    system.file("www", package = "shiny.router")
+  )
+  jsFile <- file.path("shiny.router", "shiny.router.js")
+
+  list(
     shiny::singleton(
-      shiny::tags$head(
-        shiny::tags$script(src = "shiny.router/page.js")
+      shiny::withTags(
+        head(
+          script(type = "text/javascript", src = jsFile)
+        )
       )
     ),
-    shiny::uiOutput(ROUTER_UI_ID),
-    shiny::textInput(HIDDEN_ROUTE_INPUT, label = ""),
-    shiny::tags$style(paste0("#", HIDDEN_ROUTE_INPUT, " {display: none;}"))
+    shiny::uiOutput(ROUTER_UI_ID)
   )
 }
 
+#' Convenience function to retrieve just the "page" part of the input. This
+#' corresponds to what might be called the "path" component of a URL, except
+#' that we're using URLs with hashes before the path & query (e.g.:
+#' http://www.example.com/#!/virtual/path?and=params)
+#' @param session The current Shiny Session
+#' @return The current page in a length-1 character vector, or FALSE if the input
+#' has no value.
+#' @reactivesource
+#' @export
+get_page <- function(session = shiny::getDefaultReactiveDomain()) {
+  session$userData$shiny.router.page()$path
+}
+
+#' Convenience function to retrieve any params that were part of the requested
+#' page. The param values returned come from "httr::parse_url()"
+#' @param field If provided, retrieve only a param with this name. (Otherwise,
+#' return all params)
+#' @param session The Shiny session
+#' @return The full list of params on the URL (if any), as a list. Or, the single
+#' requested param (if present). Or FALSE if there's no input, or no params.
+#' @reactivesource
+#' @export
+get_query_param <- function(field = NULL, session = shiny::getDefaultReactiveDomain()) {
+  log_msg("Trying to fetch field '", field)
+
+  if (is.null(session$userData$shiny.router.page()$query)) {
+    return(FALSE)
```

Comment:
I suppose `FALSE` output is not intuitive here. I other cases there is returned string, so maybe we could return NULL in here?

Reply:
👍